### ISSUE TITLE
Allow for specifying gem name for Ruby clients?

### ISSUE THREAD

**Issue by jgn** [swagger-api/swagger-codegen#958]
In the Ruby generator, the gem name is defined as a protected variable and cannot be set as a config option.

Is that deliberate? It seems kind of weird to be forced to generate a client called "swagger_client" and then have to do a lot of hand-editing to convert it to something more appropriate.

**Comment by wing328** [comment]
@jgn you can customize the Ruby gem using the following setting in a config.json:
```
$ java -jar modules/swagger-codegen-cli/target/swagger-codegen-cli.jar config-help -l ruby

CONFIG OPTIONS
	gemName
	    gem name (convention: underscore_case), default: swagger_client

	moduleName
	    top module name (convention: CamelCase, usually corresponding to gem name), default: SwaggerClient

	gemVersion
	    gem version, default: 1.0.0
```

For example, to generate a Ruby with the name petstore_client, I'll construct the following config.json file:
```
{
  "gemName": "petstore_client",
  "moduleName": "PetstoreClient",
  "gemVersion": "3.2.1"
}
```
and run the following command:
```
java -jar modules/swagger-codegen-cli/target/swagger-codegen-cli.jar generate \
  -i http://petstore.swagger.io/v2/swagger.json \
  -l ruby \
  -o /var/tmp/ruby/ \
  -c /var/tmp/config.json
```
Would this meet your requirement ?

**Comment by jgn** [comment]
Thanks. What branch are you on? I don't seem to have that in master.

```
$ java -jar modules/swagger-codegen-cli/target/swagger-codegen-cli.jar config-help -l ruby

CONFIG OPTIONS
	modelPackage
	    package for generated models

	apiPackage
	    package for generated api classes
```

**Comment by wing328** [comment]
Please kindly use `develop_2.0`, which contains the latest enhancements and bug fixes.

**Comment by jgn** [comment]
Got it. Thanks!

**Comment by wing328** [comment]
@jgn no problem.  Please let me know if you've any feedback or question on the Ruby API client.

**Comment by jgn** [comment]
Yes, actually. Is there an example of how to do the settings to get the `request.rb` file properly populated via mustache? I could make my own template but it seems that I should be able to get this populated through one of the options . . . (I can move this to a new issue if you like. Or maybe there's a blog post somewhere, though I haven't found it.)

```
      # Update hearder and query params based on authentication settings.
      def update_params_for_auth!
        (@auth_names || []).each do |auth_name|
          case auth_name
          {{#authMethods}}when '{{name}}'
            {{#isApiKey}}{{#isKeyInHeader}}@headers ||= {}
            @headers['{{keyParamName}}'] = get_api_key_with_prefix('{{keyParamName}}'){{/isKeyInHeader}}{{#isKeyInQuery}}@params ||= {}
            @params['{{keyParamName}}'] = get_api_key_with_prefix('{{keyParamName}}'){{/isKeyInQuery}}{{/isApiKey}}{{#isBasic}}@headers ||= {}
            http_auth_header = 'Basic ' + ["#{Swagger.configuration.username}:#{Swagger.configuration.password}"].pack('m').delete("\r\n")
            @headers['Authorization'] = http_auth_header{{/isBasic}}{{#isOAuth}}# TODO: support oauth{{/isOAuth}}
          {{/authMethods}}
          end
        end
      end
```

**Comment by wing328** [comment]
Have you tried the `-t` option to specify the template folder? Something like the following:
```
-t modules/swagger-codegen/src/main/resources/ruby
```

If you don't mind, would you please share with us how you would like to customize the `request.rb` to meet your requirement ?

**Comment by jgn** [comment]
At present, with no settings, the generator makes bad Ruby syntax. Here's what you get. A case statement with no when's doesn't parse.

```
      # Update hearder and query params based on authentication settings.
      def update_params_for_auth!
        (@auth_names || []).each do |auth_name|
          case auth_name
          
          end
        end
      end
```

And here's the template:

```
      # Update hearder and query params based on authentication settings.
      def update_params_for_auth!
        (@auth_names || []).each do |auth_name|
          case auth_name
          {{#authMethods}}when '{{name}}'
            {{#isApiKey}}{{#isKeyInHeader}}@headers ||= {}
            @headers['{{keyParamName}}'] = get_api_key_with_prefix('{{keyParamName}}'){{/isKeyInHeader}}{{#isKeyInQuery}}@params ||= {}
            @params['{{keyParamName}}'] = get_api_key_with_prefix('{{keyParamName}}'){{/isKeyInQuery}}{{/isApiKey}}{{#isBasic}}@headers ||= {}
            http_auth_header = 'Basic ' + ["#{Swagger.configuration.username}:#{Swagger.configuration.password}"].pack('m').delete("\r\n")
            @headers['Authorization'] = http_auth_header{{/isBasic}}{{#isOAuth}}# TODO: support oauth{{/isOAuth}}
          {{/authMethods}}
          end
        end
      end
```

So whoever wrote this must have been able to set ` #authMethods` (and similar settings) to make it work.

**Comment by wing328** [comment]
Ah ok. I'll look into the issue and file a PR to fix it if I manage to find a fix.

**Comment by jgn** [comment]
Cool. Any idea how those values are set during client generation?

**Comment by wing328** [comment]
As an example the [Petstore Swagger spec] (http://petstore.swagger.io/v2/swagger.json) contains `securityDefinitions`, which is used in populating those authentication-related mustache tags.

**Comment by jgn** [comment]
Superb!

On Wed, Jul 8, 2015 at 3:04 PM, wing328 <notifications@github.com> wrote:

> As an example the Petstore Swagger spec
> <http://petstore.swagger.io/v2/swagger.json> contains securityDefinitions,
> which is used in populating those authentication-related mustache tags.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/swagger-api/swagger-codegen/issues/958#issuecomment-119571131>
> .
>



-- 
John G. Norman
Chief Software Architect
Iora Health
The American Twine Building
222 Third Street, Suite 3100
Cambridge MA 02142

**Comment by xhh** [comment]
Hi @jgn, thanks for reporting the issue regarding empty authMethods. I have just pushed the fixes to PR #923 (in commit a51d71e72860ff70e84f54ce6be3c9d667a9e19c).
